src/blk/BlockDevice.cc

```diff
@@ -106,46 +106,57 @@ BlockDevice *BlockDevice::create(CephContext* cct, const string& path,
       type = "ust-nvme";
   }
 
-#if defined(HAVE_BLUESTORE_PMEM)
-  if (type == "kernel") {
-    int is_pmem = 0;
-    size_t map_len = 0;
-    void *addr = pmem_map_file(path.c_str(), 0, PMEM_FILE_EXCL, O_RDONLY, &map_len, &is_pmem);
-    if (addr != NULL) {
-      if (is_pmem)
-	type = "pmem";
-      else
-	dout(1) << path.c_str() << " isn't pmem file" << dendl;
-      pmem_unmap(addr, map_len);
-    } else {
-      dout(1) << "pmem_map_file:" << path.c_str() << " failed." << pmem_errormsg() << dendl;
+  if (front_end_type == BLUESTORE) {
```

Comment:
> blk: Limit RBD to only use KernelDevice

why?

Reply:
Currently, there're four backend drivers:
1) KernelDevice  2) PMEMDevice  3) HMSMRDevice 4) NVMEDevice
The current design of RWL-SSD is based on KernelDevice. We haven't verified it on other drivers.

Reply:
@tchaikov 
Do we need to support all the backend drivers at the same if we can't verify them? Currently, there's no plan to verify RWL-SSD's functionality on other backends except KernelDevice.

Reply:
@tchaikov ping

Reply:
@tchaikov ping

Reply:
@tchaikov ping

Reply:
@changchengx sorry for the latency. we are using a heuristic way to figure out which `BlockDevice` implementation should be created for given `path`.

i think a better way to solve this problem is to allow the caller to specify the expected backend explicitly. i'd suggest add some helper functions, like

```c++
enum class block_device_t {
  kernel,
  ust_nvme, /// probably should be "spdk" instead
  pmem,
  hm_smr,
};

// add static methods for each backends to detect if the device with given path is compatible with this backend
bool PMEMDevice:: supports(const string& path) {
  /// ...
}

block_device_t BlockDevice::detect_device_type(const string& path)
{
  if (PMEMDevice::supports(path)) {
    return block_device_t::pmem;
  } else if (...) {
  } else {
    // fall back to KernelDevice
    return block_device_t::kernel;
  }
}

BlockDevice::create(CephContext* cct, const string& path,
				 aio_callback_t cb, void *cbpriv, aio_callback_t d_cb, void *d_cbpriv)
{
  const string blk_dev_name = cct->_conf.get_val<string>("block_device_type");
  block_device_t device_type = block_device_t::kernel;
  if (blk_dev.empty()) {
    device_type = detect_device_type(path);
  } else {
    device_type = device_type_from_name(blk_dev_name);
  }
  return create_with_type(device_type, cct, path, cb, cbpriv, d_cb, d_cbpriv);
}
```

some advantages this way

1. move the detection logic down to the specific implementation, `BlockDevice` does not have to understand what its implementation expects
1. allow the caller to use specific backend on demand
1. avoid encoding the backend type into the path like `NVMEDevice` does.
1. backward compatible

what do you think?

Reply:
@tchaikov 
1. The above suggestions are to differentiate which block driver should be used to operate the block devices.
     1) User (such as client which use rbd) could set the specific block driver for the block devices with "block_device_type" configuration.
     2) If caller (such as bluestore) doesn't set "block_device_type", choose the best driver according to detect the parameter "path"(backward compatible).
      I could refine the implementation according to your sample code. 
2. This PR is to differentiate the block driver behavior according to the caller (e.g. bluestore, rbd).
     https://github.com/ceph/ceph/blob/master/src/blk/kernel/KernelDevice.cc#L69 is used to check whether using io_uring driver when the caller is bluestore. I'm worried that user may put "bluestore ioring = true" in global section which may affect the client's behavior.

Reply:
> I'm worried that user may put "bluestore ioring = true" in global section which may affect the client's behavior.

i don't think we are able to detect all combinations of configurations that not supported or tested. if you want to make sure that user does not use such settings, probably a pre-flight checklist would help. or you could check this on the caller side. but please do not encode this policy in the underlying BlockDevice.

Reply:
@tchaikov 
Pre-flight check maybe helpful, but it's not easy for the user to find that they should be put "bluestore ioring" in global section.
>please do not encode this policy in the underlying BlockDevice.
Should we remove parsing "bluestore ioring" from KernelDevice? Then, we need  ```caller transfer another argument to BlockDevice::create API to tell whether io_uring should be used or not``` However, assume the case that there're more bluestore configurations need to be parsed to affect backend driver in future, we'll need to extend BlockDevice::create to support more parameters.

Reply:
i am not sure i follow you. again, i think you could check this on the caller side.

Reply:
@tchaikov 
What's about below example to solve it in caller side?
Extend API
```cpp
BlockDevice::create(CephContext* cct, const string& path, aio_callback_t cb, void *cbpriv, aio_callback_t d_cb, void *d_cbpriv)
```
Usage
```cpp
int BlueStore::_open_bdev(bool create)
{
  bool use_ioring = cct->_conf.get_val<bool>("bluestore_ioring");
  bdev = BlockDevice::create(cct, p, aio_cb, static_cast<void*>(this), discard_cb, static_cast<void*>(this), use_ioring);
  int r = bdev->open(p);
}
```

Reply:
After talking with Kefu, we'll change "bluestore_ioring" to "bdev_ioring" to avoid giving user wrong hint info.